spirit_simulator/gazebo_scripts/src/estimator_plugin.cpp

```diff
@@ -0,0 +1,77 @@
+#include "estimator_plugin.h"
+#include <iostream>
+namespace gazebo{
+
+  SpiritEstimatorGroundTruth::SpiritEstimatorGroundTruth()
+  {
+
+  }
+  SpiritEstimatorGroundTruth::~SpiritEstimatorGroundTruth(){}
+
+  void SpiritEstimatorGroundTruth::Load(physics::ModelPtr _parent, sdf::ElementPtr _sdf)
+  {
+    model_ = _parent;
+
+    last_time_ = model_->GetWorld()->SimTime();
+    // Load update rate from SDF
+    if (_sdf->HasElement("updateRateHZ"))
+    {
+      update_rate_ =  _sdf->Get<double>("updateRateHZ");
+      ROS_INFO_STREAM("Ground Truth State Estimator: <updateRateHZ> set to: " << update_rate_);
+     }
+     else
+     {
+       update_rate_ = 100.0;
+       ROS_WARN_STREAM("Ground Truth State Estimator: missing <updateRateHZ>, set to default: " << update_rate_);
+     }
+
+    // Setup state estimate publisher
+    ros::NodeHandle nh;
+    state_estimate_pub_ = nh.advertise<spirit_msgs::StateEstimate>("/gazebo/ground_truth_state",1);
+
+      // Listen to the update event. This event is broadcast every
+      // simulation iteration.
+    updateConnection_= event::Events::ConnectWorldUpdateBegin(
+        std::bind(&SpiritEstimatorGroundTruth::OnUpdate, this));
+  }
+  void SpiritEstimatorGroundTruth::OnUpdate()
+  {
+    common::Time current_time = model_->GetWorld()->SimTime();
+    
+    if(update_rate_>0 && (current_time-last_time_).Double() < 1.0/update_rate_) //update rate check
+      return;
+
+    // Extract all relevant information from simulator
+    physics::LinkPtr body_link = model_->GetChildLink("body");
```

Comment:
Some of the other nodes look for a "dummy" link which I needed to add to the urdf to make rviz happy. We should probably make sure there isn't a discrepancy here but that can be on a later commit.

Reply:
That should be fine, the dummy link should just be rigidly attached to the body link, so long as the body link is still present in the URDF then this will still work.

Reply:
I believe the body link is rigidly attached to the dummy link, but yeah should work. Probably makes more sense for everything to reference body when possible, and only reference dummy if specifically needed.